test-data/unit/check-classes.test

```diff
@@ -616,6 +616,18 @@ main:3: error: Incompatible types in assignment (expression has type "B", variab
 main:4: error: Incompatible types in assignment (expression has type "A", variable has type "D")
 main:5: error: Incompatible types in assignment (expression has type "D2", variable has type "D")
 
+[case testClassmethodBeforeInit]
+class C:
+    @classmethod
+    def create(cls, arg: int) -> 'C':
+        return cls(arg)
+
+    def __init__(self, arg: 'int') -> None:
+        self._arg: int = arg
+
+reveal_type(C.create)  # E: Revealed type is 'def (arg: builtins.int) -> __main__.C'
+[builtins fixtures/classmethod.pyi]
```

Comment:
I am not sure this test actually tests the situation enough. I think what you need to add is a `reveal_type` for `cls` and `cls(<expected args>)` in the method body, _and_ as well check that `cls(<unexpected args>)` fails with a reasonable error.

Please be sure to include all the examples that appeared in the original issue. In addition please add at least tests that check your fix works correctly with the following:
* Fine-grained daemon
* Forward references
* Generic classes
* Overloaded methods (there are known issues, but few simple tests would help)